circuits/cpp/src/aztec3/msgpack/schema_impl.hpp

```diff
@@ -0,0 +1,190 @@
+#pragma once
+
+#include "barretenberg/common/msgpack/concepts.hpp"
+#include "barretenberg/common/msgpack_impl.hpp"
+#include "barretenberg/plonk/proof_system/verification_key/verification_key.hpp"
+#include "barretenberg/stdlib/primitives/address/address.hpp"
+#include "barretenberg/ecc/curves/bn254/fr.hpp"
+#include <memory>
+#include <concepts>
+#include <string>
+#include <array>
+#define MSGPACK_NO_BOOST
+#include "aztec3/msgpack/check_memory_span.hpp"
+#include "schema_name.hpp"
+
+namespace msgpack {
+
+/**
+ * Define a serialization schema based on compile-time information about a type being serialized.
+ * This is then consumed by typescript to make bindings.
+ */
+struct SchemaPacker : packer<msgpack::sbuffer> {
+    SchemaPacker(msgpack::sbuffer& stream)
+        : packer<msgpack::sbuffer>(stream)
+    {}
+    // For tracking emitted types
+    std::set<std::string> emitted_types;
+    // Returns if already was emitted
+    bool set_emitted(const std::string& type)
+    {
+        if (emitted_types.find(type) == emitted_types.end()) {
+            emitted_types.insert(type);
+            return false;
+        }
+        return true;
+    }
+};
+
+// Helper for packing (key, value, key, value, ...) arguments
+inline void _schema_pack_map_content(SchemaPacker&)
+{
+    // base case
+}
+// Helper for packing (key, value, key, value, ...) arguments
+template <typename Value, typename... Rest>
+inline void _schema_pack_map_content(SchemaPacker& packer, std::string key, Value value, Rest... rest)
+{
+    packer.pack(key);
+    schema_pack(packer, value);
+    _schema_pack_map_content(packer, rest...);
+}
+
+/**
+ * Schema pack base case for types with no special msgpack method.
+ * @tparam T the type.
+ * @param packer the schema packer.
+ */
+template <typename T>
+    requires(!HasMsgPack<T> && !HasMsgPackPack<T>)
+inline void schema_pack(SchemaPacker& packer, T const&)
+{
+    packer.pack(schema_name(T{}));
+}
+
+/**
+ * Pack a type indicating it is an alias of a certain msgpack type
+ * @param packer The packer.
+ * @param schema_name The CPP type.
+ * @param msgpack_name The msgpack type.
+ */
+inline void _schema_pack_alias(SchemaPacker& packer, const std::string& schema_name, const std::string& msgpack_name)
+{
+    packer.pack_array(2);
+    packer.pack("alias");
+    packer.pack_array(2);
+    packer.pack(schema_name);
+    packer.pack(msgpack_name);
```

Comment:
not following this

Reply:
I will comment, it creates ["alias", [schema_name, msgpack_name]